implementations/rust/ockam/ockam_command/src/shell/mod.rs

```diff
@@ -0,0 +1,228 @@
+            machine: None,
+            json: None,
+        }
+    }
+}
+
+/// Trait defining the main methods to write messages to the shell.
+pub trait ShellWriter: Clone {
+    fn stdout(no_color: bool) -> Self;
+    fn stderr(no_color: bool) -> Self;
+    fn is_tty(&self) -> bool;
+
+    fn write(&mut self, s: &str) -> Result<()>;
+    fn rewrite(&mut self, s: &str) -> Result<()>;
+    fn write_line(&mut self, s: &str) -> Result<()>;
+}
+
+// Core functions
+impl<W: ShellWriter> Shell<W> {
+    pub fn new(quiet: bool, no_color: bool, no_input: bool, output_format: OutputFormat) -> Self {
+        let no_color = Self::should_disable_color(no_color);
+        let no_input = Self::should_disable_user_input(no_input);
+        let stdout = W::stdout(no_color);
+        let stderr = W::stderr(no_color);
+        Self {
+            stdout,
+            stderr,
+            quiet,
+            no_input,
+            output_format,
+            mode: Logging,
+        }
+    }
+
+    pub fn can_ask_for_user_input(&self) -> bool {
+        !self.no_input && self.stderr.is_tty()
+    }
+
+    fn should_disable_color(no_color: bool) -> bool {
+        // If global argument `--no-color` is passed or the `NO_COLOR` env var is set, colors
+        // will be stripped out from output messages. Otherwise, let the terminal decide.
+        no_color || std::env::var("NO_COLOR").is_ok()
+    }
+
+    fn should_disable_user_input(no_input: bool) -> bool {
+        // If global argument `--no-input` is passed or the `NO_INPUT` env var is set we won't be able
+        // to ask the user for input.  Otherwise, let the terminal decide based on the `is_tty` value
+        no_input || std::env::var("NO_INPUT").is_ok()
+    }
+}
+
+// Logging mode
+impl<W: ShellWriter> Shell<W, Logging> {
+    pub fn write(&mut self, msg: &str) -> Result<()> {
+        if self.quiet {
+            return Ok(());
+        }
+        self.stderr.write(msg)
+    }
+
+    pub fn rewrite(&mut self, msg: &str) -> Result<()> {
+        if self.quiet {
+            return Ok(());
+        }
+        self.stderr.rewrite(msg)
+    }
+
+    pub fn write_line(&mut self, msg: &str) -> Result<()> {
+        if self.quiet {
+            return Ok(());
+        }
+        self.stderr.write_line(msg)
+    }
+
+    pub fn stdout(self) -> Shell<W, Finished> {
+        Shell {
+            stdout: self.stdout,
+            stderr: self.stderr,
+            quiet: self.quiet,
+            no_input: self.no_input,
+            output_format: self.output_format,
+            mode: Finished {
+                output: Output::new(),
+            },
+        }
+    }
+}
+
+// Finished mode
+impl<W: ShellWriter> Shell<W, Finished> {
+    pub fn human<T: Display>(mut self, msg: T) -> Self {
+        self.mode.output.human = Some(msg.to_string());
+        self
+    }
+
+    pub fn machine<T: Display>(mut self, msg: T) -> Self {
+        self.mode.output.machine = Some(msg.to_string());
+        self
+    }
```

Comment:
I think we should rename
output.human to output.normal
output.machine to output.tty

the human / machine idea is very leaky. For example json is also machine readable.

Reply:
Actually, `human` is what is shown when the terminal is interactive and `machine` when is not.
- stdout is tty -> show human-readable format
- stdout is not tty -> show machine-readable format

Context on the terms used: https://clig.dev/#output

I agree though that `machine` specifically doesn't make 100% sense, but I think it's the most accurate. By default, the format can be either `human` or `machine` and is derived from the stdout tty value. The user can override the format with the `--format` argument.

> For example json is also machine readable.

Yes, but json is a known format in itself. We can clarify the kind of output the user can expect from our `machine` format.

Reply:
Both very great points here.

I personally would see us go with plain, as opposed to normal or human.

Because of exactly what JSON/YAML, etc are.

Especially in this web based world where plain-text is often the go to vernacular for anything that isn't a standardized format that would work here.

Reply:
Renamed `human` to `plain`.